File: unused/WarehouseAppRN/src/services/StorageService.ts
import RNFS from 'react-native-fs';
import { Container } from '../types/Container';
import storage from '@react-native-firebase/storage';
import { PDFDocument, PDFPage, rgb } from 'react-native-pdf-lib';

export class StorageService {
  private tempDir = RNFS.TemporaryDirectoryPath;
  private documentsDir = RNFS.DocumentDirectoryPath;

  /**
   * Generate PDF report for a container
   */
  async generatePDF(container: Container): Promise<string> {
    try {
      const fileName = `${container.containerNumber}_report_${Date.now()}.pdf`;
      const pdfPath = `${this.documentsDir}/${fileName}`;

      // Create PDF document
      const page1 = PDFPage
        .create()
        .setMediaBox(200, 200)
        .drawText('Warehouse Container Report', {
          x: 20,
          y: 180,
          color: rgb(0, 0.2, 0.8),
          fontSize: 20,
        })
        .drawText(`Container #: ${container.containerNumber}`, {
          x: 20,
          y: 160,
          fontSize: 14,
        })
        .drawText(`Type: ${container.type}`, {
          x: 20,
          y: 145,
          fontSize: 12,
        })
        .drawText(`Door #: ${container.doorNumber || 'N/A'}`, {
          x: 20,
          y: 130,
          fontSize: 12,
        })
        .drawText(`Date: ${new Date().toLocaleDateString()}`, {
          x: 20,
          y: 115,
          fontSize: 12,
        })
        .drawText('Piece Count:', {
          x: 20,
          y: 95,
          fontSize: 14,
        });

      let yPos = 80;
      container.pieceCounts.forEach(pc => {
        page1.drawText(`  • ${pc.quantity} ${pc.packageType}`, {
          x: 20,
          y: yPos,
          fontSize: 10,
        });
        yPos -= 12;
      });

      yPos -= 10;
      page1.drawText('Materials Supplied:', {
        x: 20,
        y: yPos,
        fontSize: 14,
      });
      yPos -= 15;

      container.materialsSupplied.forEach(material => {
        page1.drawText(`  • ${material}`, {
          x: 20,
          y: yPos,
          fontSize: 10,
        });
        yPos -= 12;
      });

      if (container.discrepancies.length > 0) {
        yPos -= 10;
        page1.drawText('Discrepancies:', {
          x: 20,
          y: yPos,
          fontSize: 14,
          color: rgb(0.8, 0, 0),
        });
        yPos -= 15;

        container.discrepancies.forEach(disc => {
          page1.drawText(`  • ${disc.description}`, {
            x: 20,
            y: yPos,
            fontSize: 10,
          });
          yPos -= 12;
        });
      }

      const pdf = PDFDocument.create(pdfPath).addPages(page1);
      await pdf.write();

      return pdfPath;
    } catch (error) {
      console.error('Error generating PDF:', error);
      throw error;
    }
  }

  /**
   * Upload PDF to Firebase Storage and get shareable link
   */
  async uploadPDF(pdfPath: string, containerNumber: string): Promise<string> {
    try {
      const fileName = `${containerNumber}_${Date.now()}.pdf`;
      const reference = storage().ref(`reports/${fileName}`);

      await reference.putFile(pdfPath);
      const downloadUrl = await reference.getDownloadURL();

      return downloadUrl;
    } catch (error) {
      console.error('Error uploading PDF:', error);
      throw error;
    }
  }

  /**
   * Upload photo to Firebase Storage
   */
  async uploadPhoto(photoPath: string, containerNumber: string): Promise<string> {
    try {
      const fileName = `photo_${Date.now()}.jpg`;
      const reference = storage().ref(`containers/${containerNumber}/photos/${fileName}`);

      await reference.putFile(photoPath);
      const downloadUrl = await reference.getDownloadURL();

      return downloadUrl;
    } catch (error) {
      console.error('Error uploading photo:', error);
      throw error;
    }
  }

  /**
   * Upload multiple photos
   */
  async uploadPhotos(photoPaths: string[], containerNumber: string): Promise<string[]> {
    try {
      const uploadPromises = photoPaths.map(path => 
        this.uploadPhoto(path, containerNumber)
      );
      return await Promise.all(uploadPromises);
    } catch (error) {
      console.error('Error uploading photos:', error);
      throw error;
    }
  }

  /**
   * Delete photo from local storage
   */
  async deletePhoto(photoPath: string): Promise<void> {
    try {
      const exists = await RNFS.exists(photoPath);
      if (exists) {
        await RNFS.unlink(photoPath);
      }
    } catch (error) {
      console.error('Error deleting photo:', error);
      throw error;
    }
  }

  /**
   * Copy photo to app's document directory
   */
  async copyPhotoToDocuments(sourcePath: string): Promise<string> {
    try {
      const fileName = `photo_${Date.now()}.jpg`;
      const destPath = `${this.documentsDir}/${fileName}`;
      await RNFS.copyFile(sourcePath, destPath);
      return destPath;
    } catch (error) {
      console.error('Error copying photo:', error);
      throw error;
    }
  }

  /**
   * Get file size
   */
  async getFileSize(filePath: string): Promise<number> {
    try {
      const stat = await RNFS.stat(filePath);
      return stat.size;
    } catch (error) {
      console.error('Error getting file size:', error);
      return 0;
    }
  }

  /**
   * Clean up temporary files
   */
  async cleanupTempFiles(): Promise<void> {
    try {
      const files = await RNFS.readDir(this.tempDir);
      const deletePromises = files.map(file => RNFS.unlink(file.path));
      await Promise.all(deletePromises);
    } catch (error) {
      console.error('Error cleaning up temp files:', error);
    }
  }
}

File: lib/services/storage_service.dart
import 'dart:io';
import 'package:firebase_storage/firebase_storage.dart';
import 'package:path/path.dart' as path;

/// Service for handling file uploads to Firebase Storage
class StorageService {
  final FirebaseStorage _storage = FirebaseStorage.instance;
  
  /// Upload single photo to Firebase Storage
  Future<String> uploadPhoto({
    required String containerNumber,
    required String photoPath,
  }) async {
    try {
      final photoFile = File(photoPath);
      final fileName = '${containerNumber}_${DateTime.now().millisecondsSinceEpoch}.jpg';
      final ref = _storage.ref().child('containers/$containerNumber/photos/$fileName');
      
      final uploadTask = ref.putFile(photoFile);
      final snapshot = await uploadTask;
      final downloadUrl = await snapshot.ref.getDownloadURL();
      
      return downloadUrl;
    } catch (e) {
      throw Exception('Failed to upload photo: $e');
    }
  }

  /// Upload photos to Firebase Storage
  Future<List<String>> uploadPhotos({
    required String containerNumber,
    required List<String> photoPaths,
  }) async {
    final uploadedUrls = <String>[];
    
    try {
      for (int i = 0; i < photoPaths.length; i++) {
        final photoFile = File(photoPaths[i]);
        final fileName = '${containerNumber}_photo_${i + 1}.jpg';
        final ref = _storage.ref().child('containers/$containerNumber/photos/$fileName');
        
        final uploadTask = ref.putFile(photoFile);
        final snapshot = await uploadTask;
        final downloadUrl = await snapshot.ref.getDownloadURL();
        
        uploadedUrls.add(downloadUrl);
      }
    } catch (e) {
      throw Exception('Failed to upload photos: $e');
    }
    
    return uploadedUrls;
  }

  /// Upload PDF report to Firebase Storage
  Future<String> uploadPdfReport({
    required String containerNumber,
    required String pdfPath,
  }) async {
    try {
      final pdfFile = File(pdfPath);
      final fileName = '${containerNumber}_report.pdf';
      final ref = _storage.ref().child('containers/$containerNumber/reports/$fileName');
      
      final uploadTask = ref.putFile(pdfFile);
      final snapshot = await uploadTask;
      final downloadUrl = await snapshot.ref.getDownloadURL();
      
      return downloadUrl;
    } catch (e) {
      throw Exception('Failed to upload PDF report: $e');
    }
  }

  /// Generate a shareable folder link
  Future<String> generateShareableLink(String containerNumber) async {
    try {
      final ref = _storage.ref().child('containers/$containerNumber');
      
      // Create a signed URL that expires in 1 year
      final downloadUrl = await ref.getDownloadURL();
      
      // For a folder, we'll return the parent directory URL
      // In a real implementation, you might want to create a custom sharing mechanism
      return downloadUrl;
    } catch (e) {
      throw Exception('Failed to generate shareable link: $e');
    }
  }

  /// Delete photos from storage
  Future<void> deletePhotos({
    required String containerNumber,
    required List<String> photoUrls,
  }) async {
    try {
      for (final url in photoUrls) {
        final ref = _storage.refFromURL(url);
        await ref.delete();
      }
    } catch (e) {
      // Failed to delete some photos - continue anyway
    }
  }

  /// Delete PDF report from storage
  Future<void> deletePdfReport(String pdfUrl) async {
    try {
      final ref = _storage.refFromURL(pdfUrl);
      await ref.delete();
    } catch (e) {
      // Failed to delete PDF report
    }
  }

  /// Get temporary directory for local file operations
  Future<Directory> getTemporaryDirectory() async {
    return await getTemporaryDirectory();
  }

  /// Get application documents directory
  Future<Directory> getApplicationDocumentsDirectory() async {
    return await getApplicationDocumentsDirectory();
  }

  /// Create a local file path for temporary storage
  Future<String> createTempFilePath(String fileName) async {
    final tempDir = await getTemporaryDirectory();
    return path.join(tempDir.path, fileName);
  }

  /// Create a local file path in documents directory
  Future<String> createDocumentsFilePath(String fileName) async {
    final docsDir = await getApplicationDocumentsDirectory();
    return path.join(docsDir.path, fileName);
  }

  /// Check if file exists locally
  bool fileExists(String filePath) {
    return File(filePath).existsSync();
  }

  /// Get file size in bytes
  Future<int> getFileSize(String filePath) async {
    final file = File(filePath);
    return await file.length();
  }

  /// Copy file to a new location
  Future<String> copyFile(String sourcePath, String destinationPath) async {
    final sourceFile = File(sourcePath);
    final copiedFile = await sourceFile.copy(destinationPath);
    return copiedFile.path;
  }

  /// Move file to a new location
  Future<String> moveFile(String sourcePath, String destinationPath) async {
    final sourceFile = File(sourcePath);
    final movedFile = await sourceFile.rename(destinationPath);
    return movedFile.path;
  }

  /// Delete local file
  Future<void> deleteLocalFile(String filePath) async {
    final file = File(filePath);
    if (await file.exists()) {
      await file.delete();
    }
  }

  /// Clean up temporary files
  Future<void> cleanupTempFiles() async {
    try {
      final tempDir = await getTemporaryDirectory();
      final files = tempDir.listSync();
      
      for (final file in files) {
        if (file is File) {
          // Delete files older than 24 hours
          final stat = await file.stat();
          final now = DateTime.now();
          final fileAge = now.difference(stat.modified);
          
          if (fileAge.inHours > 24) {
            await file.delete();
          }
        }
      }
    } catch (e) {
      // Failed to cleanup temp files
    }
  }
}
